### DISCUSSION TITLE
Convert real to float with @UseQt true

### DISCUSSION THREAD

**Discussion by SRaude** [thelfer/tfel#354] [General]
I tried to use @UseQt keyword in my MFront file.
In my MFront file, I first declared a material property which is called mB, with the type real: @MaterialProperty real mB;
Then, I wanted to be sure that mB is an integer.
Without @UseQt keyword, I just declared: const int mBi=mB; or const int mBi=static_cast<int>(mB); and then use mBi instead of mB in the @Integrator block ; and it's fine ; but with @UseQt true, it didn't work.
Is there a solution to convert a real to an integer with @UseQt true ? Or to use a type to declare an integer as a @MaterialProperty ?

Simon

**Comment by maxence-wz** [comment]
@SRaude the `UseQt` keyword allows to check the principle of homogeneity of dimensions in your equations. I think this is not what you are trying to do. Your solution seems good :)

**Reply by SRaude** [reply]
@maxence-wz, thank you for your quick answer!

Actually, my problem is that the compilation of the MFront (mfront-4.2.0-dev --obuild --interface=aster test.mfront) file is not OK when the @UseQt keyword is uncommented but OK in the other case. 

I've tried to reproduce the case in a simple example in the following.

```
@DSL Implicit;
// @UseQt true;
@Behaviour Test;
@MaterialProperty real young;

@InitLocalVariables{
	const int youngi=young;
}

@Integrator{
	const int youngi=young;
}
```

**Reply by thelfer** [reply]
@SRaude In this case, you can use the `base_type_cast` function as follows:

```cxx
@InitLocalVariables{
	const int youngi=base_type_cast(young);
}
```

If young is a quantity (i.e. if `UseQt` is `true`), then it calls the `getValue` method to access the underlying floatting point value. Otherwise, the value is just forwarded. Hence this code works in both cases :)

**Reply by SRaude** [reply]
@thelfer, thank you for your quick answer !
@thelfer and @maxence-wz, I also take this opportunity to wish you a happy new year ;-)
See you soon,
Simon

**Reply by SRaude** [reply]
@thelfer, your solution seems to work perfectly, thank you again!
Last question: is there a solution to declare some quantity (real, strain, stress,...) to the power n (which is not something simple like 2, 3, 4)?

**Reply by thelfer** [reply]
> Last question: is there a solution to declare some quantity (real, strain, stress,...) to the power n (which is not something simple like 2, 3, 4)?

No, its does not make any sense :)

More precisely, such operations generally hides a normalisation factor that must be explicit when using quantities.

For exemple, one shall never use `pow(sigeq, n)` but `pow(sigeq/sig0, n)` (note that `sigeq/sig0` is dimensionless, so it is implicitely convertible to the underlying floating point type, hence you can call standard mathematical functions).

**Comment by thelfer** [comment]
@SRaude Happy new year too !
